Lombiq.Tests.UI/ui-testing-toolkit.mjs

```diff
@@ -0,0 +1,132 @@
+}
+
+async function _takeScreenshot(driver, file){
+    let image = await driver.takeScreenshot()
+    await writeFile(file, image, 'base64')
+}
+
+/**
+ * Verifies that the provided element's inner text contains the provided text.
+ * @param {WebElement} element The web element whose inner text is examined.
+ * @param {string} text The expected inner text fragment.
+ * @returns {Promise<void>} Success when the element text contains the expected string, rejection if it does not or if
+ *                          element is null or empty.
+ */
+async function shouldContainText(element, text) {
+    if (element?.then) {
+        element = await element;
+    }
+
+    if (!element) {
+        throw new Error('The element is missing.');
+    }
+
+    const actualText = (await element.getText())?.trim();
+
+    if (actualText?.includes(text) !== true) {
+        const url = await element.getDriver().getCurrentUrl();
+        throw new Error(
+            `Expected element at ${url} to contain text "${text}", but it does not. (Actual text: ${actualText})`);
+    }
+}
+
+/**
+ * Navigates the browser to the given URL and then verifies that the page has loaded with non-empty content.
+ * @param {WebDriver} driver The driver whose current tab should be navigated.
+ * @param {string} url The target URL.
+ * @param {number} maxAttempts The maximum number of attempts. If exceeded, an error is thrown.
+ * @returns {Promise<void>} Success when a non-empty page has been reached.
+ */
+async function navigate(driver, url, maxAttempts = 10) {
+    for (let i = 0; i < maxAttempts; i++) {
+        await driver.navigate().to(url);
+
+        await driver.wait(() => driver
+            .executeScript('return document.readyState')
+            .then((readyState) => readyState === 'complete'));
+
+        try {
+            if ((await driver.findElement(By.xpath('//body')).getAttribute('innerHTML'))?.trim()) {
+                console.log(`Successfully reached ${url}.`);
+                return;
+            }
+        }
+        catch (exception) {
+            // Nothing to do here, let's try again.
+        }
+    }
+
+    throw new Error(`Failed to navigate to a non-empty page at ${url} in ${maxAttempts} attempts.`)
+}
+
+/**
+ * Executes a test by connecting to an existing web driver using the information in the command line arguments.
+ * @param {function(WebDriver, string):Promise<void>} test
+ * @param {function(chrome.Options):chrome.Options} configureOptions Update the configuration before the driver is built.
+ * @returns {Promise<void>} Success if the driver is created and the test has run to completion.
+ */
+async function runTest(test, configureOptions = null) {
+    const args = process.argv.slice(2);
+    if (args.length !== 3) throw new Error('Usage: node script.js driverPath startUrl tempDirectory');
+    const [driverPath, startUrl, tempDirectory] = args;
+
+    let options = new chrome.Options()
```

Comment:
Doesn't this mean that the JS test-related C# methods should throw an exception if the current browser is not Chrome, because this won't work?

Reply:
No, because they can provide their own driver config function if they want to. But I've updated the code to throw about it more explicitly in JS.